File: C++/bee2308.cpp
#include <bits/stdc++.h>

#define endl '\n'
#define _ ios_base::sync_with_stdio(0);cin.tie(0);cout.tie(NULL);
#define dbg(a) (cout << #a << " = " << a << '\n')
#define loop(i,a,m) for(int i = a; i < m; ++i)
#define pb push_back
#define f first
#define s second

using namespace std;

typedef uint32_t uint;
typedef long long ll;
typedef long double ld;
typedef vector<int> vtr;
typedef pair<int,int> par;
typedef vector<vector<par>> gph;

const int INF = 0x3f3f3f3f;
const ll LINF = 0x3f3f3f3f3f3f3f3fll;

int dijkstra(int i, gph &g) {
    priority_queue<par> q; q.push({0, i});
    vtr vis(g.size(), 0);
    int path{INF};

    while (!q.empty()) {
        par p = q.top(); q.pop();
        vis[p.s] = 1;

        for(auto &v : g[p.s]) {
            if(v.s == i) path = min(path, -(p.f - v.f));

            if(!vis[v.s]) q.push({p.f - v.f, v.s});
        }
    }

    return path;
}

int main() {_
    int s, c; cin >> s >> c;
    vtr tempos(s);
    gph salas(s);
    loop(i, 0, s) cin >> tempos[i];

    loop(i, 0, c) {
        int a, b, c; cin >> a >> b >> c;
        a--; b--;
        salas[a].pb({c + tempos[a], b});
    }

    int r{INF};
    loop(i, 0, s) r = min(r, dijkstra(i, salas));

    cout << r << endl;

    return 0;
}

File: Python/bee2308.py
import heapq

INF = float('inf')

def dijkstra(i, g):
    vis = [False] * len(g)
    path = INF
    q = []
    heapq.heappush(q, (0, i))
    
    while len(q) > 0:
        a, b = heapq.heappop(q)
        vis[b] = True

        for d, v in g[b]:
            if v == i:
                path = min(path, a + d)

            if not vis[v]:
                heapq.heappush(q, (a + d, v))

    return path

def main():
    s, c = map(int, input().split())
    tempos = list(map(int, input().split()))
    salas = [[] for _ in range(s)]

    for i in range(c):
        x, y, z = map(int, input().split())
        salas[x-1].append((z + tempos[x-1], y-1))

    r = INF
    for i in range(s):
        r = min(r, dijkstra(i, salas))

    print(r)    

if __name__ == '__main__':
    main()